RecFCCeeCalorimeter/src/components/CaloTowerToolFCCee.cpp

```diff
@@ -0,0 +1,432 @@
+#include "CaloTowerToolFCCee.h"
+
+// FCCSW
+#include "k4Interface/IGeoSvc.h"
+
+// datamodel
+#include "edm4hep/CalorimeterHitCollection.h"
+#include "edm4hep/Cluster.h"
+#include "edm4hep/MutableCluster.h"
+
+// DD4hep
+#include "DD4hep/Detector.h"
+#include "DD4hep/Readout.h"
+
+DECLARE_COMPONENT(CaloTowerToolFCCee)
+
+CaloTowerToolFCCee::CaloTowerToolFCCee(const std::string& type, const std::string& name, const IInterface* parent)
+    : GaudiTool(type, name, parent), m_geoSvc("GeoSvc", name) {
+  declareProperty("ecalBarrelCells", m_ecalBarrelCells, "");
+  declareProperty("ecalEndcapCells", m_ecalEndcapCells, "");
+  declareProperty("ecalFwdCells", m_ecalFwdCells, "");
+  declareProperty("hcalBarrelCells", m_hcalBarrelCells, "");
+  declareProperty("hcalExtBarrelCells", m_hcalExtBarrelCells, "");
+  declareProperty("hcalEndcapCells", m_hcalEndcapCells, "");
+  declareProperty("hcalFwdCells", m_hcalFwdCells, "");
+  declareInterface<ITowerToolThetaModule>(this);
+}
+
+StatusCode CaloTowerToolFCCee::initialize() {
+  if (GaudiTool::initialize().isFailure()) {
+    return StatusCode::FAILURE;
+  }
+  
+  if (!m_geoSvc) {
+    error() << "Unable to locate Geometry Service. "
+            << "Make sure you have GeoSvc and SimSvc in the right order in the configuration." << endmsg;
+    return StatusCode::FAILURE;
+  }
+
+  // check if readouts exist & retrieve Module-Theta segmentations
+  // if readout does not exist, reconstruction without this calorimeter part will be performed
+  std::pair<dd4hep::DDSegmentation::Segmentation*, SegmentationType> tmpPair;
+  info() << "Retrieving Ecal barrel segmentation" << endmsg;
+  tmpPair = retrieveSegmentation(m_ecalBarrelReadoutName);
+  m_ecalBarrelSegmentation = tmpPair.first;
+  m_ecalBarrelSegmentationType = tmpPair.second;
+  if (tmpPair.first != nullptr && tmpPair.second == SegmentationType::kWrong) {
```

Comment:
Here, you are checking for the existing segmentation with wrong type. Can you also check for the case when you have non existent segmentation and also wrong type?
```
if (tmpPair.first == nullptr && tmpPair.second == SegmentationType::kWrong) {
```
This kind of pair is returned in two cases from the function `retrieveSegmentation()`.

Reply:
ok, made changes like:
```
  if (tmpPair.second == SegmentationType::kWrong) {
    if (tmpPair.first != nullptr) {
      error() << "Wrong type of segmentation" << endmsg;
    } else {
      error() << "Segmentation does not exist" << endmsg;
    }
    return StatusCode::FAILURE;
  }
```

Reply:
Sorry, now I understand. These check are there to determine whether you have certain segmentation type available. So there should not be return  `StatusCode::FAILURE`. I'm not sure if the check is needed at all since you are always returning `nullptr` for the non-valid segmentation. Can you in the `retrieveSegmentation()` function instead of `error()` log level use `warning()`?

Reply:
ok ```error()``` in ```retrieveSegmentation()``` now changed to ```warning()```.